[clang-format] Fix erroneous BraceWrapping.BeforeLambdaBody column calcs

jamesg-nz opened the pull request:
Firstly, must check ColumnLimit > 0 before comparing calculated columns to the limit. Otherwise it always breaks before the brace if ColumnLimit = 0 (no limit). Fixes #50275

Secondly, the lambda body length alone is currently compared with the column limit. Should instead be comparing a column position, which includes everything before the lambda body, the body length itself, and any unbreakable tail. Fixes #59724

Thirdly, if must break before the lambda right brace, e.g. line comment in body, then must also break before the left brace. Can't use column calculation in this instance.

jamesg-nz committed:
[clang-format] Fix erroneous BraceWrapping.BeforeLambdaBody column calcs

Firstly, must check ColumnLimit > 0 before comparing calculated columns
to the limit. Otherwise it always breaks before the brace if ColumnLimit
= 0 (no limit). Fixes #50275

Secondly, the lambda body length alone is currently compared with the
column limit. Should instead be comparing a column position, which
includes everything before the lambda body, the body length itself, and
any unbreakable tail. Fixes #59724

Thirdly, if must break before the lambda right brace, e.g. line comment
in body, then must also break before the left brace. Can't use column
calculation in this instance.

jamesg-nz commented:
n.b. there's relevant test coverage provided by existing `LambdaWithLineComments` test.

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -22965,6 +22965,84 @@ TEST_F(FormatTest, EmptyLinesInLambdas) {
+               "    });",
+               Style);
+
+  for (unsigned l : {0, 41}) {
```
Don't loop. Pick some values.

jamesg-nz replied:
Now testing ColumnLimit = 0 once separately. Unrolled remaining loop (that doesn't test = 0) into two verifications.

Loops make it hard to see in test results for what value it is failing for, I guess. Although saw other people had used them in that file.

HazardyKnusperkeks replied:
They make it hard to see which is failing, yes. But also why introduce additional runtime for the tests, if there is no real gain?

jamesg-nz committed:
Per feedback, remove loops from test.

Have one separate check for ColumnLimit = 0 case.

Unroll one loop (that doesn't use 0) into two verifications.

jamesg-nz commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -366,8 +366,14 @@ bool ContinuationIndenter::mustBreak(const LineState &State) {
       Current.is(TT_LambdaLBrace) && Previous.isNot(TT_LineComment)) {
-    auto LambdaBodyLength = getLengthToMatchingParen(Current, State.Stack);
-    return LambdaBodyLength > getColumnLimit(State);
+    if (Current.MatchingParen->MustBreakBefore)
```
I noticed there is code in `TokenAnnotator` than can set `MustBreakBefore` on the lambda left brace token:
https://github.com/llvm/llvm-project/blob/0e01c72c5645259d9a08a1a7ed39cb5cc41ce311/clang/lib/Format/TokenAnnotator.cpp#L5297-L5299
Introduced in 37c2233097ac44697b87228d86eef1fce10ea5c1 seemingly for code like this:

```cpp
auto select = [this]() -> std::unique_ptr<Object> { return MyAssignment::SelectFromList(this); }
```

Both before this current PR and here the `MustBreakBefore` state of the token is ignored. Should behavior change to observe this and thus force a break?

Would seem the _BraceWrapping.BeforeLambdaBody_ setting would then force lambda across multiple lines if on, but not if off.

Should the code in `TokenAnnotator` be changed and maybe only set `CanBreakBefore`?

HazardyKnusperkeks replied:
> I noticed there is code in `TokenAnnotator` than can set `MustBreakBefore` on the lambda left brace token:
> 
> https://github.com/llvm/llvm-project/blob/0e01c72c5645259d9a08a1a7ed39cb5cc41ce311/clang/lib/Format/TokenAnnotator.cpp#L5297-L5299
> 
> Introduced in [37c2233](https://github.com/llvm/llvm-project/commit/37c2233097ac44697b87228d86eef1fce10ea5c1) seemingly for code like this:
> 
> ```c++
> auto select = [this]() -> std::unique_ptr<Object> { return MyAssignment::SelectFromList(this); }
> ```
> 
> Both before this current PR and here the `MustBreakBefore` field of the `Current` token is ignored. Should behavior change to observe this and thus force a break?
> 
> Would seem the _BraceWrapping.BeforeLambdaBody_ setting would then force lambda across multiple lines if on, but not if off.
> 
> Should the code in `TokenAnnotator` be changed and maybe only set `CanBreakBefore`?

So as far as I put everything together we should break before the `{`, if `BeforeLambdaBody` is set. Except for short lambdas, if `AllowShortLambdasOnASingleLine` matches our lambda **and** it would fit on a single line. If the lambda has to be broken somewhere, it should break before `{`.

What is needed to be changed I don't know, but also have not the time to investigate. I'm willing to hear your analysis, or maybe @mydeveloperday can shed some light into the issue?

jamesg-nz replied:
@HazardyKnusperkeks - sorry late response. After looking into it, I don't believe there's an issue.

The code I was talking about above: https://github.com/llvm/llvm-project/blob/0e01c72c5645259d9a08a1a7ed39cb5cc41ce311/clang/lib/Format/TokenAnnotator.cpp#L5297-L5299

Results in `MustBreakBefore` being set for the token (the left lambda brace), which therefore results in `CanBreakBefore` being set. The code exists as there's conditions (e.g. [`Left.is(TT_PointerOrReference)`](https://github.com/llvm/llvm-project/blob/0e01c72c5645259d9a08a1a7ed39cb5cc41ce311/clang/lib/Format/TokenAnnotator.cpp#L5523-L5524)) in `TokenAnnotator::canBreakBefore()` which return `false` for the left lambda brace in lambdas like these:

```c++
auto select = [this]() -> const Library::Object * { return MyAssignment::SelectFromList(this); };
auto select = [this]() -> const Library::Object & { return MyAssignment::SelectFromList(this); };
auto select = [this]() -> std::unique_ptr<Object> { return MyAssignment::SelectFromList(this); };"
```

It's the fact it sets `CanBreakBefore` that matters so it enters the lambda section in `ContinuationIndenter::mustBreak()`.

Just checking `Current.MatchingParen->MustBreakBefore` inside is sufficient as TokenAnnotator::mustBreakBefore() returns `true` when it is called for the right lambda:
https://github.com/llvm/llvm-project/blob/0e01c72c5645259d9a08a1a7ed39cb5cc41ce311/clang/lib/Format/TokenAnnotator.cpp#L5283-L5293

Thus there is no need to look at `Current.MustBreakBefore`. However the generic condition that checks `Current.MustBreakBefore` for all tokens is in the condition for the next `if()` statement down, so could move the lambda `if()` statement down beneath that. But there's no need to do this at the moment.

So this PR looks OK as-is with respect to this.

jamesg-nz committed:
Merge branch 'main' into jg-beforelambdabody-calc-fixes

jamesg-nz committed:
Merge branch 'main' into jg-beforelambdabody-calc-fixes

owenca commented:
Ping @mydeveloperday
